### PR TITLE
Dealing with optional and mandatory GATK resource files

### LINKED ISSUE #299: Error in GetPileupSummaries  when using --no_intervals option
I got this error message

`2020-10-23T12:17:09.403+02:00	A USER ERROR has occurred: Argument intervals was missing: Argument 'intervals' is required.`

The command executed was:

```
Command executed:
  gatk --java-options "-Xmx7g"         GetPileupSummaries         -I 3109.recal.bam         -V gnomAD.r2.1.1.GRCh38.PASS.AC.AF.only.vcf.gz                  -O no_intervals_3109_pileupsummaries.table
But I can see in the GATK documentation that the argument -L --intervals is required.
```

But I can see in the GATK documentation that the argument -L --intervals is required.

### PR DESCRIPTION
**Pull request by FriederikeHanssen**
Closes #299  , #359 , #367 

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by FriederikeHanssen** [comment]
@maxulysse still unsure on how to deal with `known_sites` (see nextflow slack). For the rest, would love your opinion. I am not super happy with prepare_genome.nf, however I coudln't find another work around. If you know of some nextflow magic there, i would be all 👂

**Comment by FriederikeHanssen** [comment]
@maxulysse carefully testing now all combinations again, but so far so good.

**Review by maxulysse — APPROVED** [review]
<3

**Comment by maxulysse** [comment]
Can you do the channel.value magic for all of the reference files?

**Comment by FriederikeHanssen** [comment]
yes should be able to, the only exception is germline_resource_tbi because I need it to check later whether getpileup summaries should be run. I couldn't get naything else to work  with`Channel.value([])`, tried `Channel.ifEmpty` , `germline_resource_tbi ?:` etc.

**Review by maxulysse — APPROVED** [review]

**Comment by FriederikeHanssen** [comment]
Not sure what is going on with mutect2. Locally results files are there:
```
results
├── csv
│   └── variantcalled.csv
├── multiqc
│   ├── multiqc_data
│   │   ├── mqc_bcftools-stats-subtypes_1.txt
│   │   ├── mqc_bcftools_stats_depth_1.txt
│   │   ├── mqc_bcftools_stats_indel-lengths_1.txt
│   │   ├── mqc_bcftools_stats_vqc_Count_Indels.txt
│   │   ├── mqc_bcftools_stats_vqc_Count_SNP.txt
│   │   ├── mqc_bcftools_stats_vqc_Count_Transitions.txt
│   │   ├── mqc_bcftools_stats_vqc_Count_Transversions.txt
│   │   ├── mqc_vcftools_tstv_by_count_1.txt
│   │   ├── mqc_vcftools_tstv_by_qual_1.txt
│   │   ├── multiqc.log
│   │   ├── multiqc_bcftools_stats.txt
│   │   ├── multiqc_citations.txt
│   │   ├── multiqc_data.json
│   │   ├── multiqc_general_stats.txt
│   │   ├── multiqc_sources.txt
│   │   ├── vcftools_tstv_by_count.txt
│   │   └── vcftools_tstv_by_qual.txt
│   ├── multiqc_plots
│   │   ├── pdf
│   │   │   ├── mqc_bcftools-stats-subtypes_1.pdf
│   │   │   ├── mqc_bcftools-stats-subtypes_1_pc.pdf
│   │   │   ├── mqc_bcftools_stats_depth_1.pdf
│   │   │   ├── mqc_bcftools_stats_indel-lengths_1.pdf
│   │   │   ├── mqc_bcftools_stats_vqc_Count_Indels.pdf
│   │   │   ├── mqc_bcftools_stats_vqc_Count_SNP.pdf
│   │   │   ├── mqc_bcftools_stats_vqc_Count_Transitions.pdf
│   │   │   ├── mqc_bcftools_stats_vqc_Count_Transversions.pdf
│   │   │   ├── mqc_vcftools_tstv_by_count_1.pdf
│   │   │   └── mqc_vcftools_tstv_by_qual_1.pdf
│   │   ├── png
│   │   │   ├── mqc_bcftools-stats-subtypes_1.png
│   │   │   ├── mqc_bcftools-stats-subtypes_1_pc.png
│   │   │   ├── mqc_bcftools_stats_depth_1.png
│   │   │   ├── mqc_bcftools_stats_indel-lengths_1.png
│   │   │   ├── mqc_bcftools_stats_vqc_Count_Indels.png
│   │   │   ├── mqc_bcftools_stats_vqc_Count_SNP.png
│   │   │   ├── mqc_bcftools_stats_vqc_Count_Transitions.png
│   │   │   ├── mqc_bcftools_stats_vqc_Count_Transversions.png
│   │   │   ├── mqc_vcftools_tstv_by_count_1.png
│   │   │   └── mqc_vcftools_tstv_by_qual_1.png
│   │   └── svg
│   │       ├── mqc_bcftools-stats-subtypes_1.svg
│   │       ├── mqc_bcftools-stats-subtypes_1_pc.svg
│   │       ├── mqc_bcftools_stats_depth_1.svg
│   │       ├── mqc_bcftools_stats_indel-lengths_1.svg
│   │       ├── mqc_bcftools_stats_vqc_Count_Indels.svg
│   │       ├── mqc_bcftools_stats_vqc_Count_SNP.svg
│   │       ├── mqc_bcftools_stats_vqc_Count_Transitions.svg
│   │       ├── mqc_bcftools_stats_vqc_Count_Transversions.svg
│   │       ├── mqc_vcftools_tstv_by_count_1.svg
│   │       └── mqc_vcftools_tstv_by_qual_1.svg
│   └── multiqc_report.html
├── pipeline_info
│   ├── execution_report_2022-06-20_08-48-17.html
│   ├── execution_timeline_2022-06-20_08-48-17.html
│   ├── execution_trace_2022-06-20_08-48-17.txt
│   ├── pipeline_dag_2022-06-20_08-48-17.html
│   └── software_versions.yml
├── reports
│   ├── bcftools
│   │   └── sample4_vs_sample3.filtered.bcftools_stats.txt
│   └── vcftools
│       ├── sample4_vs_sample3.filtered.FILTER.summary
│       ├── sample4_vs_sample3.filtered.TsTv.count
│       └── sample4_vs_sample3.filtered.TsTv.qual
├── untar
│   └── chromosomes
│       └── chr21.fasta
└── variant_calling
    └── sample4_vs_sample3
        └── mutect2
            ├── sample3.pileupsummaries.table
            ├── sample4.pileupsummaries.table
            ├── sample4_vs_sample3.artifactprior.tar.gz
            ├── sample4_vs_sample3.contamination.table
            ├── sample4_vs_sample3.filtered.vcf.gz
            ├── sample4_vs_sample3.filtered.vcf.gz.filteringStats.tsv
            ├── sample4_vs_sample3.filtered.vcf.gz.tbi
            ├── sample4_vs_sample3.segmentation.table
            ├── sample4_vs_sample3.vcf.gz
            ├── sample4_vs_sample3.vcf.gz.stats
            └── sample4_vs_sample3.vcf.gz.tbi
```

**Review by maxulysse — APPROVED** [review]

File: conf/modules.config [hunk 1]
```diff
@@ -846,13 +846,22 @@ process{
         ]
     }
 
-    withName: 'GETPILEUPSUMMARIES.*' {
+     withName: 'GETPILEUPSUMMARIES.*' {
```
Comment (maxulysse) [inline conf/modules.config:849]:
```suggestion
    withName: 'GETPILEUPSUMMARIES.*' {
```

Comment (maxulysse) [reply conf/modules.config:849]: I think you had the right indentation before already

File: conf/modules.config [hunk 2]
```diff
@@ -846,13 +846,22 @@ process{
...
         publishDir        = [
             mode: params.publish_dir_mode,
-            path: { "${params.outdir}/variant_calling/${meta.id}/" },
+            path: { "${params.outdir}/variant_calling/" },
+            pattern: "*.table",
+            saveAs: { meta.num_intervals > 1 ? null : "${meta.id}/mutect2/${it}" }
+        ]
+    }
+
+    withName: 'GETPILEUPSUMMARIES_.*' {
```
Comment (maxulysse) [inline conf/modules.config:859]: why is there this selector twice?

Comment (FriederikeHanssen) [reply conf/modules.config:859]: once with `_` (for `NORMAL` , `TUMOR` in paired mutect2 usage ) and the one without for all GETPILEUPSUMMARIES. Just added that to fix something :D

Comment (FriederikeHanssen) [reply conf/modules.config:859]: hopefully I didn't break more stuff in the process 🐘 🍽️

Comment (maxulysse) [reply conf/modules.config:859]: still same selector, the `saveAs` is overwritten

Comment (FriederikeHanssen) [reply conf/modules.config:859]: yes but that is the point 🤔 for paired it should publish into ` "${meta.tumor_id}_vs_${meta.normal_id}/mutect2/${it}"` and for single it should publish into `"${meta.id}/mutect2/${it}"`

Comment (maxulysse) [reply conf/modules.config:859]: :facepalm: 
Neverming, I got it, I saw the `...PILEUPSUMMARIES_.*` twice and on top of that did not pay enough attention to the `_` part...

**Comment by FriederikeHanssen** [comment]
locally it didn't work with 7.5GB of memory. I am afraid that mutect test just won't run on GHA

**Comment by maxulysse** [comment]
> locally it didn't work with 7.5GB of memory. I am afraid that mutect test just won't run on GHA

Too bad, then, let's run that test locally only whenever needed.
I'm assuming we already have input data as small as possible?

**Comment by FriederikeHanssen** [comment]
yep Gavin downsampled it as much as he could but the somatic GATK tools require a certain amount of SNPs to run at all. If we further reduce they'll fail for other reasons and much more upstream

**Pull request merged**